tape390/tape390_display.8
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH TAPE390_DISPLAY 8 "Apr 2006" "s390-tools"

.SH NAME
tape390_display \- display messages on the display unit of a zSeries tape
device.

.SH SYNOPSIS
.TP 16
.B tape390_display
.RB [ \-b | \-\-blink ]
.br
.RB [ \-h | \-\-help ]
.br
.RB [ \-l | \-\-load ]
.br
.RB [ \-t <type>| \-\-type <type>]
.br
.RB [ \-q | \-\-quiet ]
.br
.RB [ \-v | \-\-version ]
.br
.IR message1
.RB [
.I message2
.RB ]
.I device

.SH DESCRIPTION
\fBtape390_display\fR displays one or two 8-byte messages on the display
unit of a zSeries tape device.  Characters to be displayed include capital
letters A-Z, numerics 0-9 and special characters.
However some special characters may not display on all device types. The
characters that should display on model 3490 are:
.br
	'@$#,./()*&+-=%|:_<>?;' and the <cent> character.
.br
Spaces and shell-sensitive characters have to be quoted.

.SH OPTIONS
.TP 8
\fB\-h\fR or \fB\-\-help\fR
Print help text.

.TP
\fB\-b\fR or \fB\-\-blink\fR
Display \fImessage1\fR repeatedly for 2 seconds at half-second intervals.
This option is ignored if two messages are specified on command line.

.TP
\fB\-l\fR or \fB\-\-load\fR
If this option is given the display request will cause a load request to be
sent to the automatic tape loader (if installed). The next indexed tape will
then be loaded if the loader is in 'system' mode. Otherwise the request is
ignored.

.TP
\fB\-q\fR or \fB\-\-quiet\fR
This option causes the command to run in quiet mode which will suppress
all warning messages.

.TP
\fB\-t\fR \fItype\fR or \fB\-\-type\fR \fItype\fR
Set the type of the message to be displayed. This influences how long
the message(s) stay on the display. Available types are:

.RS
.TP 16
.B standard
The message(s) stay displayed until the drive executes the next tape
movement command. This is the default.
.TP
.B load
Displays the message(s) until a tape is loaded. If there already is
a tape loaded to the device, nothing is displayed at all.
.TP
.B unload
This will display the message(s) until currently loaded tape is unloaded.
Again, if the drive \fIis\fR currently unloaded there is no output at all.
.TP
.B noop
For testing purposes only, this will access the drive but won't do anything
at all (no operation). This doesn't seem to work with 3590 tape devices.
.TP
.B reload
This display type \fBrequires\fR two messages. The first message is
displayed as long as there is a tape loaded up to the time the unload
operation is completed (or, if there isn't a tape loaded, it will be
ignored). The second message is displayed as soon as a tape gets loaded
into the drive.
.RE

.TP
\fB\-v\fR or \fB\-\-version\fR
Print the S/390 tools package version, the utility version and exit.

.TP 8
\fBmessage1\fR [ \fBmessage2\fR ]
The message(s) to be displayed. They stay displayed until either the
ending condition for the message type takes place or the display is
updated with another message.
.br
If two messages are specified, they are alternated on the display.
Each message is then displayed repeatedly for 2 seconds
at half-second intervals.

.TP
\fBdevice\fR
The tape device on which the message(s) will be displayed, e.g.:
.sp
	\fB/dev/ntibm\fR\fI0\fR
.br

.SH EXAMPLES
\fBtape390_display \-\-blink \-\-type unload "TESTING" /dev/ntibm\fR\fI0\fR
.RS
Let the message TESTING blink on the display of the tape unit until the
tape is removed (or another message is given). If there is no tape in the
drive, nothing happens.
.RE
\fBtape390_display "BACKUP" "COMPLETE" /dev/ntibm\fR\fI0\fR
.RS
This will display the messages BACKUP and COMPLETE each for 2 seconds
until the tape is moved, ejected or (if there isn't a tape in the
device) inserted.
.RE
\fBtape390_display \-\-type reload "UNLOAD" "LOAD" /dev/ntibm\fR\fI0\fR
.RS
The UNLOAD message will be displayed if there is a tape inserted and as
long as the tape isn't removed. After that the LOAD message is displayed
until a tape is inserted. If there is no tape inserted when the command
.RE
\fBtape390_display \-\-load \-\-type load "LOADING" /dev/ntibm\fR\fI0\fR
.RS
Causes the device to display the message LOADING until a tape is inserted
and to instruct the automatic stacker to load the next tape (if the mode
of operation is \fIsystem\fR). If there already is a tape loaded, nothing
will happen at all.
.RE

.SH AUTHOR
.nf
This man-page was written by Despina Papadopoulou<despina_p@de.ibm.com>
.nf
Updates by Stefan Bader<shbader@de.ibm.com>
.fi



CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (2023-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.32.0 (2024-04-03)__

  For Linux kernel version: 6.8

  Changes of existing tools:
  - cpumf/lscpumf: add support for machine type 3932
  - genprotimg, pvattest, and pvsecret accept IBM signing key with Armonk as
    subject locality
  - zdump/zipl: Support for List-Directed dump from ECKD DASD
  - zkey: Detect FIPS mode and generate PBKDF for luksFormat according to it

  Bug Fixes:
  - dbginfo.sh: dash compatible copy sequence
  - rust/pv_core: Fix UvDeviceInfo::get() method
  - zipl/src: Fix leak of files if run with a broken configuration
  - zkey: Fix convert command to accept only keys of type CCA-AESDATA

* __v2.31.0 (2024-02-02)__

  For Linux kernel version: 6.7

  General:
  - common.mak: Set default C/C++ standard to gnu11/gnu++11

  Add new tools / libraries:
  - pvapconfig: Tool to automatically configure APQNs in SE KVM guests
  - s390-tools: Provide pre-commit configuration

  Changes of existing tools:
  - cpuplugd: Adjust to CPU 0 being no longer hotpluggable
  - dbginfo.sh: Check for Dynamic Partition Mode
  - dbginfo.sh: Update man page and copyright
  - rust/pv: Add user-data signing and verifying
  - rust/pvsecret: Add user defined signatures and verifications
  - zdev/dracut: Consolidate device configuration

  Bug Fixes:
  - dbginfo.sh: Fix relative path on script copy
  - libkmipclient: Fix build with libxml2-2.12.0
  - pvsecret: Fix panic if empty file is used as host key document
  - rust/pv: Fix 'elided_lifetimes_in_associated_constant' warning

* __v2.30.0 (2023-12-01)__

  For Linux kernel version: 6.6

  Add new tools / libraries:
  - lspai: Tool to display PAI counter sets
  - s390-tools: Provide a ShellCheck configuration

  Changes of existing tools / libraries:
  - cpumf/pai: Add command line option for realtime scheduling
  - dbginfo.sh: enhance ethtool collection for ROCE
  - libutil/util_lockfile: add routine to return owning pid of file lock
  - lszcrypt: Improve lszcrypt output on SE guests
  - rust: Use a single workspace for all rust tools
  - zdev: limit the derivation of ZDEV_SITE_ID
  - zdump/df_s390: Update 'zgetdump -i' output with zlib info
  - zdump/dfi_s390: Support reading compressed s390_ext dumps
  - zipl/boot: Integrate zlib compression to single volume DASD dumper
  - zipl/boot: compile the bootloaders only if HOST_ARCH is s390x
  - zipl: Add --no-compress option to zipl command
  - zkey: Also check for deconfigured and check-stopped cards
  - dbginfo.sh: fix relative path on script copy

  Bug Fixes:
  - ap_tools/ap-check: handle get-attributes between pre and post event
  - libutil: fix util_file_read_*() using wrong format specifiers
  - rust/pv: fix Invalid write of size 1

* __v2.29.0 (2023-08-04)__

  For Linux kernel version: 6.5

  General:
  - s390-tools now supports tools written in Rust.
  - Add `compdb` Makefile target to create 'compile_commands.json' to LSP
    backends in IDEs and editors

  Add new tools / libraries:
  - rust/pv: Library for pv tools written in rust
  - rust/pvsecret: Tool to manage UV-secrets

  Changes of existing tools:
  - dbginfo.sh: Global IFS variable
  - genprotimg: Add support for add-secret requests
  - genprotimg: Build debuginfo files for bootloader
  - hyptop: Add real SMT utilization field
  - hyptop: Allow users to set speedup factor
  - pvattest: Add yaml-output for verify command
  - zipl: Build debuginfo files for bootloader

  Bug Fixes:
  - dump2tar: Fix truncated paths
  - zdev/dracut: fix kdump build to integrate with site support

* __v2.28.0 (2023-07-11)__

  For Linux kernel version: 6.4

  Changes of existing tools:
  - chzcrypt: Support for SE AP pass-through support
  - genprotimg: Add support for non-s390x architectures
  - lszcrypt: Support for SE AP pass-through support
  - zdev: Add support for autoquiesce related sysfs attributes

  Bug Fixes:
  - ap_tools/ap-check: Handle missing 'matrix' and 'control_domains' attrs
  - ap_tools/ap-check: Hold ap config file lock over get attributes
  - s390-tools: Fix build for ppc64le
  - zdev: Add missing label in the udev-rules
  - zdev: Add proper value input for the ZDEV_SITE_ID key
  - zdev: Use rename-file to avoid any symlinks created
  - zipl/dump: fix ngdump dracut helper script

* __v2.27.0 (2023-05-30)__

  For Linux kernel version: 6.3

  Changes of existing tools:
   - s390-tools cross-compile and non-s390x support:
      - `pkg-config` is now mandatory for the build process
      - Add `PKG_CONFIG` Makefile variable to select pkg-config program;
        default `pkg-config` or `$(CROSS_COMPILE)pkg-config` if
        `CROSS_COMPILE` is set
      - Rename Makefile variable `ARCH` to `HOST_ARCH`. `HOST_ARCH` is the
        architecture that will run the produced (executable) objects
      - Add the Makefile variable `BUILD_ARCH`. `BUILD_ARCH` is the
        architecture of the build system. For each Makefile variable like
        `CC`, `LINK`, `CPP`, ... there is a suffixed version of it - e.g.
        `CC_FOR_BUILD`. This is useful for cross compiling, and this naming
        convention is very similar to the Meson convention (see
        https://mesonbuild.com/Reference-tables.html#environment-variables-per-machine).
      - Limit build targets for non-s390x architectures (pvattest)
  - dasdfmt: Fall back to full format if space release fails
  - dbginfo.sh: Add nstat for network and SNMP stats
  - dbginfo.sh: Rework crypto data collection
  - hyptop: Show thread util by default
  - zipl: Add support for list-directed IPL dump from ECKD DASD

  Bug Fixes:
  - lszcrypt: Fix argument parsing
  - zdev/dracut: Fix out-of-memory (OOM) situations in the kdump crashkernel environment
  - ziomon/ziorep_config: Fix for SCSI devices of type disk without block dev

tape390/tape390_display.c
/*
 * tape390_display - Display messages on the display unit of a tape drive
 *
 * Copyright IBM Corp. 2002, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <ctype.h>
#include <getopt.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include "lib/zt_common.h"
#include "tape390_common.h"

#define TAPE390_DISPLAY _IOW('d', 1, struct display_struct)

/*
 * The TAPE390_DISPLAY belongs to linux-2.5/include/asm-s390/tape390.h
 *
 * The TAPE390_DISPLAY ioctl calls the Load Display command
 * which transfers 17 bytes of data from the channel to the subsystem:
 *     - 1 format control byte, and
 *     - two 8-byte messages
 *
 * Format control byte:
 *   0-2: New Message Overlay
 *     3: Alternate Messages
 *     4: Blink Message
 *     5: Display Low/High Message
 *     6: Reserved
 *     7: Automatic Load Request
 *
 */
#define TAPE_MSGTYPE_STANDARD           0
#define TAPE_MSGTYPE_UNLOAD             1
#define TAPE_MSGTYPE_LOAD               2
#define TAPE_MSGTYPE_NOOP               3
#define TAPE_MSGTYPE_UNLOAD_LOAD        7

typedef struct display_struct {
	struct {
		unsigned char   type            :3;
		unsigned char   alternate       :1;
		unsigned char   blink           :1;
		unsigned char   use_second      :1;
		unsigned char   reserved        :1;
		unsigned char   load_request    :1;
	}                       cntrl;
	char                    message1[8];
	char                    message2[8];
} __attribute__ ((packed))      display_struct;

/*  definitions  */
static char *help_text =
	"tape390_display: display one or two 8-byte messages on the display"
	"unit of a tape drive\n"
	"\n"
	"Usage: tape390_display <options> \"<message1>\" [\"<message2>\"] "
	"<device>"
	"\n\n"
	"where <options> are:\n"
	"\t-b | --blink\n"
	"\t\twill cause a single message to blink every 2 seconds\n"
	"\t\t(this option has only effect with a single message)\n"
	"\t-h | --help\n"
	"\t\tprint this text\n"
	"\t-v | --version\n"
	"\t\toutput version information and exit\n"
	"\t-l | --load\n"
	"\t\twill try to load the next tape if the loader is in system mode\n"
	"\t-q | --quiet\n"
	"\t\twill suppress all warning messages\n"
	"\t-t | --type \"standard|load|unload|noop|reload\"\n"
	"\t\tcontrols how the message is displayed:\n"
	"\t\t\tstandard = until the next tape movement (default)\n"
	"\t\t\tload     = until a tape is loaded\n"
	"\t\t\tunload   = until a tape is unloaded\n"
	"\t\t\tnoop     = not at all (test purposes)\n"
	"\t\t\treload   = message1 until tape is unloaded and message2\n"
	"\t\t\t           when the next tape is loaded\n"
	"\n"
	"Note: Characters to be displayed include capital letters (A-Z), "
	"numerics (0-9)\n"
	"and special characters within single quotes. However not all "
	"characters\n"
	"might be displayed on all tape devices. Special characters that "
	"are supported\n"
	"for 3490 are '@$#,./()*&+-=%|:_<>?;'.\n";

static int quiet;

/* end of definitions */

static int typename2int(char *name)
{
	char *typenames[] = {
		"standard",
		"unload",
		"load",
		"noop",
		NULL,
		NULL,
		NULL,
		"reload"
	};
	int	i;

	for(i = 0; i < 8; i++) {
		if(typenames[i] == NULL)
			continue;
		if(strcmp(typenames[i], name) == 0) {
			return i;
		}
	}

	fprintf(stderr, "Invalid message type <%s>\n", name);
	return -1;
}

static void strchkcpy(char *tgt, const char *src)
{
	static	char *nowarn = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
			       "@$#,./'()*&+-=%:_<>?; ";
	char	c;
	char	warned;

	warned = 0;
	for(c=0; c<8 && *(src+c); c++) {
		*(tgt+c) = toupper(*(src+c));
		if (index(nowarn, *(tgt+c)) == NULL && !quiet && !warned) {
			fprintf(
				stderr,
				"WARNING: Some special characters may not "
				"display on all device types.\n"
			);
			warned = 1;
		}
	}
}

int main(int argc, char *argv[]) {

	char			*options = "hblqt:v";
	struct option		 long_options[] = {
		{ "help",	 no_argument,		NULL, 'h' },
		{ "blink",	 no_argument,		NULL, 'b' },
		{ "load",	 no_argument,		NULL, 'l' },
		{ "quiet",	 no_argument,		NULL, 'q' },
		{ "type",	 required_argument,	NULL, 't' },
		{ "version",     no_argument,           NULL, 'v' },
		{ NULL,		 0,			NULL, 0   }
	};
	char			*pathname;
	int			 c;
	int			 fd, ret = 0;
	struct display_struct	 ds;

	/* set name of program */
	set_prog_name(basename(argv[0]));

	memset(&ds, 0, sizeof(ds));
	ds.cntrl.type = TAPE_MSGTYPE_STANDARD;

	while(
		(c = getopt_long(argc, argv, options, long_options, NULL))
		!= EOF
	) {
		switch(c) {
			case 'h':
				fprintf(stdout, "%s", help_text);
				exit(0);
			case 'b':
				ds.cntrl.blink = 1;
				break;
			case 'l':
				ds.cntrl.load_request = 1;
				break;
			case 'q':
				quiet = 1;
				break;
			case 't': {
				int type = typename2int(optarg);

				if(type < 0)
					exit(1);

				ds.cntrl.type = type;
				break;
			}
			case 'v': {
				printf("tape390_display: zSeries tape "
				       "display control program version %s\n",
					RELEASE_STRING);
				printf("Copyright IBM Corp. 2002, 2017\n");
				exit(0);
			}
			default:
				fprintf(stderr, "Try '%s --help' for more"
						" information.\n",prog_name);
				exit(1);
		}
	}

	if(optind + 2 < argc ) {
		if(
			ds.cntrl.type == TAPE_MSGTYPE_LOAD &&
			ds.cntrl.load_request
		) {
			if(!quiet) {
				fprintf(
					stderr,
					"WARNING: "
					"A <load> message with the "
					"--load option will only "
					"display the first"
					"\nmessage.\n"
				);
			}
		} else if(
			ds.cntrl.type == TAPE_MSGTYPE_UNLOAD_LOAD &&
			ds.cntrl.load_request
		) {
			if(!quiet) {
				fprintf(
					stderr,
					"WARNING: "
					"A <reload> message with the "
					"--load option will only "
					"display the second"
					"\nmessage.\n"
				);
			}
		} else {
			ds.cntrl.alternate = 1;

			if(ds.cntrl.blink) {
				ds.cntrl.blink = 0;
				if(!quiet) {
					fprintf(
						stderr,
						"Alternate messages "
						"override blinking\n"
					);
				}
			}
		}
		strchkcpy(ds.message1, argv[optind]);
		strchkcpy(ds.message2, argv[optind+1]);
		pathname = argv[optind+2];
	} else if(optind + 1 < argc) {
		if(ds.cntrl.type == TAPE_MSGTYPE_UNLOAD_LOAD) {
			fprintf(stderr, "Reload message type requires "
					"two messages.\n");
			exit(1);
		} else if(ds.cntrl.type == TAPE_MSGTYPE_LOAD) {
			strchkcpy(ds.message2, argv[optind]);
		}
		strchkcpy(ds.message1, argv[optind]);
		pathname = argv[optind+1];
	} else {
		fprintf(stderr, "%s", help_text);
		exit(1);
	}

	/* check whether specified device node is tape */
	if (is_not_tape(pathname))
		exit(EXIT_MISUSE);

	/* open device                                 */
	fd = open_tape(pathname);

	ret = ioctl(fd, TAPE390_DISPLAY, &ds);
	if (ret != 0)
		fprintf(stderr, "TAPE390_DISPLAY failed\n");

	close(fd);

	return ret;
}

